### PR TITLE
Add wallet migration flow

### LINKED ISSUE #517: Legacy Wallet Migration Flow
## Summary
Add legacy wallet migration UX around `walletLoader().migrateWallet`.

## Complexity
- `L`

## Dependencies
- `10`
- `12`
- `02`

## Global Constraints
- All implementation lives under `gui-qml/`.
- No GUI runtime dependency on `bitcoin/src/qt` components. Reused logic must be moved/ported into `gui-qml/`.
- Do not change `interfaces::Wallet` or `interfaces::Node` in this version.
- Issue completion must satisfy `DefinitionOfDone.md`.

## Scope
- Detect migration-eligible wallets.
- Add migration flow with passphrase prompt where needed.
- Display migration output, backup path, and follow-up actions.
- Handle partial migration and rollback messaging.

## Acceptance Criteria (DoD-Aligned)
1. Eligible wallet can be migrated from UI.
2. Failure and partial warning conditions are clearly explained.
3. Migration backup location is shown to user.

## Required Tests
- C++ unit tests: Unit tests for migration result mapping and edge-case handling.
- QML QTest: QTests for migration confirmation, password, and result states.
- Python E2E (`qml/test/testbridge.h`): E2E migration scenario with success + forced failure path checks.

## Implementation Notes
- Document limitations in feature note because design coverage is currently not explicit.

### LINKED ISSUE #696: Remove brittle string error checking from detecting if wallet migration is needed
The current qml wallet model uses string matching on error text to see if wallet migration is needed.

Instead, we should use the listwalletdir function, get its info return, which tells us if the wallet is bdb.

### PR DESCRIPTION
**Pull request by johnny9**
02-wallet-selector
03-migration-prompt
04-migration-complete

Fixes #517 
Fixes #696

### CODE REVIEW

**Comment by johnny9** [comment]
Related to #517

File: qml/pages/wallet/ImportWalletMigration.qml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            objectName: "walletMigrationActionButton"
+            Layout.preferredWidth: Math.min(335, parent.width - 2 * Layout.leftMargin)
+            Layout.topMargin: 30
+            Layout.leftMargin: 20
+            Layout.rightMargin: 20
+            Layout.alignment: Qt.AlignCenter
+            enabled: !root.updatingPhase
+            text: {
+                if (root.successPhase) {
+                    return qsTr("Next")
```
Comment (MarnixCroes) [inline qml/pages/wallet/ImportWalletMigration.qml]:
```suggestion
                    return qsTr("Done")
```
makes more sense

File: qml/pages/wallet/ImportWalletMigration.qml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            visible: !root.updatingPhase
+            enabled: visible
+            iconSource: "image://images/caret-left"
+            text: qsTr("Back")
+            onClicked: {
+                walletController.clearWalletMigrationStatus()
+                root.back()
+            }
+        }
+        rightItem: NavButton {
```
Comment (MarnixCroes) [inline qml/pages/wallet/ImportWalletMigration.qml]: Having both `cancel` and `back` is unnecesary

File: qml/pages/wallet/ImportWalletMigration.qml [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        }
+
+        Header {
+            Layout.topMargin: 18
+            Layout.fillWidth: true
+            Layout.leftMargin: 20
+            Layout.rightMargin: 20
+            header: {
+                if (root.successPhase) {
+                    return qsTr("Update completed")
```
Comment (MarnixCroes) [inline qml/pages/wallet/ImportWalletMigration.qml]: nit, maybe `Wallet successfully migrated` is better?

**Review by MarnixCroes — COMMENTED** [review]
f0fe7f758d9a4c32e8e2780a10d27770a9d90494

Maybe you're aware, but I don't see it mentioned, there should be a way to enter a passphrase if needed.

**Comment by johnny9** [comment]
> f0fe7f758d9a4c32e8e2780a10d27770a9d90494
> 
> Maybe you're aware, but I don't see it mentioned, there should be a way to enter a passphrase if needed.

548 adds the password dialog. So I guess depending on which gets merged first I'll add the dialog for migration instead of a blank password.

**Review by MarnixCroes — APPROVED** [review]
ACK 96b06c31429c35a66292908922ba381ae69e5680

**Comment by johnny9** [comment]
from [96b06c3](https://github.com/bitcoin-core/gui-qml/commit/96b06c31429c35a66292908922ba381ae69e5680) to [1a86cdf](https://github.com/bitcoin-core/gui-qml/commit/1a86cdf11714ace60a809882484498c7e7135641)
 - trivial rebase

**Comment by johnny9** [comment]
another trivial rebase

**Comment by johnny9** [comment]
from [3ecc865](https://github.com/bitcoin-core/gui-qml/commit/3ecc86572ff77d89bd92de7ffd8ee7ad88ddbc24) to [5055da4](https://github.com/bitcoin-core/gui-qml/commit/5055da406e3f1c9498a07f93f3c24e78f9f8e03f)
- fix the duplicate test property  `objectName: "walletSelectList"` that was also added in recently merged PR
- added checkpoint/screenshots to the qml test

**Comment by johnny9** [comment]
Isolated QSettings configs in the qml functional tests as the recent Settings PR that changes languages leaked into other tests which resulted in expectations of labels not matching anymore

**Review by jarolrod — COMMENTED** [review]
ACK 596fc76df48afc9830079d7a8c904cba4e2ef26d

As a gui wallet flow, for unencrypted legacy wallets

But the detection of legacy should be reworked, we currently do brittle string error matching to detect if migration is needed. We shouldn't be doing this and instead we should be using `m_node.walletLoader().listWalletDir()` and querying it's `info` return to see if the wallet is BDB, then expose this to qml.

As the wallet controller is not being updated in this PR, this can be done in a separate PR.

For example, the following patch should convey the idea (patch doesn't touch your string matching code)

<details>

```diff
diff --git a/qml/models/walletlistmodel.cpp b/qml/models/walletlistmodel.cpp
index def378b7d..ead1e793b 100644
--- a/qml/models/walletlistmodel.cpp
+++ b/qml/models/walletlistmodel.cpp
@@ -6,8 +6,6 @@
 
 #include <interfaces/node.h>
 
-#include <QSet>
-
 WalletListModel::WalletListModel(interfaces::Node& node, QObject *parent)
 : QAbstractListModel(parent)
 , m_node(node)
@@ -16,17 +14,25 @@ WalletListModel::WalletListModel(interfaces::Node& node, QObject *parent)
 
 void WalletListModel::listWalletDir()
 {
-    QSet<QString> existing_names;
-    for (int i = 0; i < rowCount(); ++i) {
-        QModelIndex index = this->index(i, 0);
-        QString name = data(index, NameRole).toString();
-        existing_names.insert(name);
-    }
-
-    for (const auto& [path, info] : m_node.walletLoader().listWalletDir()) {
+    for (const auto& [path, format] : m_node.walletLoader().listWalletDir()) {
         QString qname = QString::fromStdString(path);
-        if (!existing_names.contains(qname)) {
-            addItem({ qname });
+        QString qformat = QString::fromStdString(format);
+        bool found = false;
+        for (int i = 0; i < rowCount(); ++i) {
+            if (m_items[i].name != qname) {
+                continue;
+            }
+            found = true;
+            if (m_items[i].format != qformat) {
+                m_items[i].format = qformat;
+                const QModelIndex changed_index = index(i, 0);
+                Q_EMIT dataChanged(changed_index, changed_index, {FormatRole, NeedsMigrationRole});
+            }
+            break;
+        }
+
+        if (!found) {
+            addItem({ qname, qformat });
         }
     }
 }
@@ -47,6 +53,10 @@ QVariant WalletListModel::data(const QModelIndex &index, int role) const
     case Qt::DisplayRole:
     case NameRole:
         return item.name;
+    case FormatRole:
+        return item.format;
+    case NeedsMigrationRole:
+        return item.format == QStringLiteral("bdb");
     default:
         return QVariant();
     }
@@ -56,6 +66,8 @@ QHash<int, QByteArray> WalletListModel::roleNames() const
 {
     QHash<int, QByteArray> roles;
     roles[NameRole] = "name";
+    roles[FormatRole] = "format";
+    roles[NeedsMigrationRole] = "needsMigration";
     return roles;
 }
 
diff --git a/qml/models/walletlistmodel.h b/qml/models/walletlistmodel.h
index c36cb2b4f..13ae99037 100644
--- a/qml/models/walletlistmodel.h
+++ b/qml/models/walletlistmodel.h
@@ -8,6 +8,7 @@
 #include <interfaces/wallet.h>
 #include <QAbstractListModel>
 #include <QList>
+#include <QString>
 
 namespace interfaces {
 class Node;
@@ -22,7 +23,9 @@ public:
     ~WalletListModel() = default;
 
     enum Roles {
-        NameRole = Qt::UserRole + 1
+        NameRole = Qt::UserRole + 1,
+        FormatRole,
+        NeedsMigrationRole,
     };
 
     int rowCount(const QModelIndex &parent = QModelIndex()) const override;
@@ -35,6 +38,7 @@ public Q_SLOTS:
 private:
     struct Item {
         QString name;
+        QString format;
     };
 
     void addItem(const Item &item);
diff --git a/qml/pages/wallet/DesktopWallets.qml b/qml/pages/wallet/DesktopWallets.qml
index fd3a488bc..8beb671ed 100644
--- a/qml/pages/wallet/DesktopWallets.qml
+++ b/qml/pages/wallet/DesktopWallets.qml
@@ -88,6 +88,9 @@ Page {
                 onAddWallet: {
                     root.addWallet()
                 }
+                onWalletMigrationRequested: (walletPath) => {
+                    root.handleWalletMigrationRequired(walletPath)
+                }
             }
         }
         centerItem: RowLayout {
diff --git a/qml/pages/wallet/WalletSelect.qml b/qml/pages/wallet/WalletSelect.qml
index b525f4c6f..f40a3673f 100644
--- a/qml/pages/wallet/WalletSelect.qml
+++ b/qml/pages/wallet/WalletSelect.qml
@@ -18,6 +18,7 @@ Popup {
     clip: true
 
     signal addWallet()
+    signal walletMigrationRequested(string walletPath)
 
     background: Item {
         anchors.fill: parent
@@ -78,17 +79,22 @@ Popup {
 
             delegate: WalletBadge {
                 required property string name;
+                required property bool needsMigration;
                 objectName: "walletSelectItem_" + name
 
                 width: 220
                 height: 32
-                text: name
-                checked: walletController.selectedWallet.name == name
+                text: needsMigration ? qsTr("%1 (needs migration)").arg(name) : name
+                checked: !needsMigration && walletController.selectedWallet.name == name
                 ButtonGroup.group: buttonGroup
                 showBalance: false
                 showIcon: false
                 onClicked: {
-                    walletController.setSelectedWallet(name)
+                    if (needsMigration) {
+                        root.walletMigrationRequested(name)
+                    } else {
+                        walletController.setSelectedWallet(name)
+                    }
                     root.close()
                 }
             }
diff --git a/test/qml/qml_tests_main.cpp b/test/qml/qml_tests_main.cpp
index b4fcc9426..2bcfb7d4d 100644
--- a/test/qml/qml_tests_main.cpp
+++ b/test/qml/qml_tests_main.cpp
@@ -961,7 +961,9 @@ class MockWalletListModel : public QAbstractListModel
 
 public:
     enum Roles {
-        NameRole = Qt::UserRole + 1
+        NameRole = Qt::UserRole + 1,
+        FormatRole,
+        NeedsMigrationRole,
     };
 
     int rowCount(const QModelIndex& parent = QModelIndex{}) const override
@@ -974,12 +976,18 @@ public:
     {
         if (!index.isValid() || index.row() < 0 || index.row() >= m_wallet_names.size()) return {};
         if (role == NameRole) return m_wallet_names.at(index.row());
+        if (role == FormatRole) return QStringLiteral("sqlite");
+        if (role == NeedsMigrationRole) return false;
         return {};
     }
 
     QHash<int, QByteArray> roleNames() const override
     {
-        return {{NameRole, "name"}};
+        return {
+            {NameRole, "name"},
+            {FormatRole, "format"},
+            {NeedsMigrationRole, "needsMigration"},
+        };
     }
 
     Q_INVOKABLE void listWalletDir() {}
```

</details>

**Comment by johnny9** [comment]
I went ahead and reworked the flow to use the format string from the listWalletDir() call. I wasn't aware that the second string in the returned pair was actually the format and that could be used to decide on if migration is needed. This makes it way less fragile of a decision than trying to interpret error strings that may change.

Unit tests added to validate and original migration flow functional test is passing

**Review by MarnixCroes — APPROVED** [review]
tested ACK 09ba06cbc35b27ca94ad1c1240f7538638f059b9

**Comment by jarolrod** [comment]
ACK https://github.com/bitcoin-core/gui-qml/commit/09ba06cbc35b27ca94ad1c1240f7538638f059b9

Thanks for reworking our migration detection, the patch accomplishes what I said we should do [here](https://github.com/bitcoin-core/gui-qml/pull/539#pullrequestreview-4287111858)

I would make the description and title explicit that this is for unencrypted only.

File: qml/pages/wallet/ImportWalletMigration.qml [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    required property string walletPath
+    signal back
+    signal cancel
+    signal next
+
+    property string phase: "intro"
+    readonly property bool introPhase: phase === "intro"
+    readonly property bool updatingPhase: phase === "updating"
+    readonly property bool successPhase: phase === "success"
+    readonly property bool failedPhase: phase === "failed"
```
Comment (pseudoramdom) [inline qml/pages/wallet/ImportWalletMigration.qml]: Can this be an enum since all of them are mutually exclusive states?

Comment (johnny9) [reply qml/pages/wallet/ImportWalletMigration.qml]: can switch this to using states

File: qml/walletqmlcontroller.cpp
Comment (pseudoramdom) [inline qml/walletqmlcontroller.cpp:722]: I ran in to this while trying to migrate wallet with passphrase.
Image

Does #548 cover this use case?

Comment (jarolrod) [reply qml/walletqmlcontroller.cpp:722]: yeah this PR should make it explicit that its setting this up, and it currently only works for unencrypted.

Password PR is intended to then finish this feature

File: qml/pages/wallet/ImportWalletMigration.qml (hunk 4 above)
Comment (johnny9) [reply qml/pages/wallet/ImportWalletMigration.qml]: i will switch to the state pattern as it is more organized but the enums dont provide value since we dont need type checking and they are functionally equivalent to strings

**Comment by pseudoramdom** [comment]
ACK b17fe2eead02acf51424ed759c1f076753b03077 (migration for for unencrypted wallets)

**Review by pseudoramdom — APPROVED** [review]

**Comment by jarolrod** [comment]
ACK https://github.com/bitcoin-core/gui-qml/commit/b17fe2eead02acf51424ed759c1f076753b03077

Changes since my last review is refactoring for states
